docs/proposals/20231122-automate-aks-features.md

```diff
@@ -0,0 +1,311 @@
+template can be applied to all Clusters built from that ClusterClass.
+
+This method makes all ManagedCluster fields available to define in a template which could lead to
+misconfiguration if certain parameters that must be unique to a cluster are erroneously shared through a
+template. Since those fields cannot be automatically identified and may evolve between AKS API versions, CAPZ
+will not attempt to categorize ASO ManagedCluster fields that way like it does between the fields present and
+omitted from the `AzureClusterClassSpec` type, for example. CAPZ could document a best-effort list of known
+fields which could or should not be defined in template types and will otherwise rely on AKS to provide
+reasonable error messages for misconfigurations.
+
+Like [Option 1], this method keeps particular versions of CAPZ decoupled from particular API versions of ASO
+resources (including allowing preview versions) and opens the door for streamlined adoption of existing AKS
+clusters.
+
+#### Option 3: CAPZ resource defines an entire unstructured ASO resource inline
+
+This method is functionally equivalent to [Option 2] except that the template resource is defined inline
+within the AzureManagedControlPlane:
+
+```go
+type AzureManagedControlPlaneClassSpec struct {
+	...
+	
+	// ManagedClusterTemplate is the ASO ManagedCluster to be used as a template from which new
+	// ManagedClusters will be created.
+	ManagedClusterTemplate map[string]interface{} `json:"managedClusterTemplate"`
+}
+```
+
+One variant of this method could be to using `string` instead of `map[string]interface{}` for the template
+type, though that would make defining patches unwieldy (like for ClusterClass).
+
+Compared to [Option 2], this method loses schema and webhook validation that would be performed by ASO when
+creating a separate ManagedCluster to serve as a template. That validation would still be performed when CAPZ
+creates the ManagedCluster resource, but that would be some time after the AzureManagedControlPlane is created
+and error messages may not be quite as visible.
+
+#### Option 4: CAPZ resource defines an entire typed ASO resource inline
+
+This method is functionally equivalent to [Option 3] except that the template field's type is the exact same
+as an ASO ManagedCluster:
+
+```go
+type AzureManagedControlPlaneClassSpec struct {
+	...
+	
+	// ManagedClusterSpec defines the spec of the ASO ManagedCluster managed by this AzureManagedControlPlane.
+	ManagedClusterSpec v1api20230201.ManagedCluster_Spec `json:"managedClusterTemplate"`
+}
+```
+
+This method allows CAPZ to leverage schema validation defined by ASO's CRDs upon AzureManagedControlPlane
+creation, but would still lose any further webhook validation done by ASO unless CAPZ can invoke that itself.
+
+It also has the drawback that one version of CAPZ is tied directly to a single AKS API version. The spec could
```

Comment:
The upside of that is that it still allows for testing before adopting new versions. ASO is not a Cluster API "provider" so we won't always be able guarantee backwards compatibility between all ASO versions and all CAPZ versions.

Reply:
I agree with this upside but I also wonder though if it'll end up in a situation where bumping the version number becomes cumbersome and potentially falls far behind. Is this a concern with this approach?

Reply:
@mtougeron I don't see that being much different than the current state of affairs where we need to manually bump the AKS API version that CAPZ uses internally. I think between maintainers staying on top of things as we upgrade ASO, users requesting features that require new API versions, and e2e tests, there's very little risk we'd release a version of CAPZ using an AKS API version so old that it doesn't function.

Reply:
> we'd release a version of CAPZ using an AKS API version so old that it doesn't function.

I'm not concerned about that. I don't foresee that being an issue either.

I was more curious about whether or not we'd end up in a state where it becomes difficult to upgrade the version like it is today of the go sdk. While there is the "buyer beware" option of modifying non-CAPZ supported feature, the happy path of using CAPZ will be tied to these version bumps.

Reply:
Can you expand on what makes updating the version difficult with the SDK? Currently those are automated dependabot PRs to grab the new Go module version for a particular API which haven't required any extra work recently that I can remember.

Reply:
For example, we were limited to what version of the `github.com/Azure/azure-sdk-for-go/services/containerservice/mgmt/2022-03-01/containerservice` that it could be updated to in https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2838. I just wasn't sure if the pinning of the version in the struct could create the same sort of problem. If it doesn't, then great! I really don't know enough about the ASO details to know if it does or not which is why I asked the question.

Reply:
Sorry, I'm not sure I remember what that issue was. Is there a specific thread in that PR where that was discussed? I couldn't find any after a quick scan.

Reply:
I think that might have been related to https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/3409

That was a one time migration because the Azure SDK itself had a new major release with breaking changes, and new API versions for AKS weren't part of v1 that we were still using. This should not happen in the future for "regular" SDK version bumps, it should be as easy as changing the go module version.